File: App/Model/Account/Account.m

#import "Account.h"
#import "API.h"
#import "MBApp.h"
#import "AccountEntity.h"
#import "Common.h"

@interface Account ()
#if MBUserStringUID
@property (readwrite) MBIdentifier uid;
#else
@property (readwrite) MBID uid;
#endif
@end

@implementation Account
@dynamic uid;

+ (void)onCurrentUserChanged:(Account *)user {
    NSUserDefaults *ud = AppUserDefaultsShared();
    ud.lastUserID = user.uid;
    ud.userToken = user.token;
    ud.AccountEntity = user.information.toJSONString;
    BOOL success = ud.synchronize;
    if (!success) {
        dispatch_after_seconds(0, ^{
            if (AppUserDefaultsShared().synchronize) return;
            
            DebugLog(YES, @"UDSynchronizeFail", @"用户信息存储失败");
            static dispatch_once_t onceToken;
            dispatch_once(&onceToken, ^{
                UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"系统错误" message:@"暂时不能保存您的用户信息，如果你反复遇到这个提示，建议您重启设备以解决这个问题" preferredStyle:UIAlertControllerStyleAlert];
                [alert addAction:[UIAlertAction actionWithTitle:@"本次启动不再提示" style:UIAlertActionStyleDefault handler:nil]];
                [UIViewController.rootViewControllerWhichCanPresentModalViewController presentViewController:alert animated:YES completion:nil];
            });
        });
    }
    if (user) {
        [AppEnv() setFlagOn:MBENVFlagUserHasLogged];
    }
    else {
        [AppEnv() setFlagOff:MBENVFlagUserHasLogged];
        [AppEnv() setFlagOff:MBENVFlagUserInfoFetched];
    }
}

- (void)onLogin {
    dout_info(@"当前用户 ID: %@, token: %@", @(self.uid), self.token);
    AppAPI().defineManager.authorizationHeader[@"token"] = self.token;
}

- (void)onLogout {
    [AppAPI().defineManager.authorizationHeader removeObjectForKey:@"token"];
    [self resetCookies];
    [self.profile synchronize];
}

#pragma mark - init

+ (void)setup {
    if (AppUser()) return;

    NSUserDefaults *ud = AppUserDefaultsShared();
    if (!ud.lastUserID) return;

    Account *user = [[self alloc] initWithID:ud.lastUserID];
    if (!user.token.length) {
        DebugLog(YES, @"LaunchUserNoToken", @"Account has ID but no token");
        return;
    }
    [self setCurrentUser:user];
    
    [API backgroundRequestWithName:@"AccountRefresh" parameters:nil completion:^(BOOL success, id  _Nullable responseObject, NSError * _Nullable error) {
        if (![responseObject isKindOfClass:NSDictionary.class]) {
            return;
        }
        NSString *newToken = responseObject[@"token"];
        if (![newToken isKindOfClass:NSString.class]
            || !user.isCurrent) return;
        if (![user.token isEqualToString:newToken]) {
            user.token = newToken;
            [user onLogin];
            AppUserDefaultsShared().userToken = newToken;
        }
        [user updateInformationFromViewController:nil complation:nil];
    }];
    RFAssert(API.global, @"");
}

- (void)onInit {
    [super onInit];
    BOOL debugServer = NO;
#if MBUserStringUID
    NSString *suitName = [NSString stringWithFormat:@"User%@%@", self.uid, debugServer? @"D" : @""].rf_MD5String;
#else
    NSString *suitName = [NSString stringWithFormat:@"User%lld%@", self.uid, debugServer? @"D" : @""].rf_MD5String;
#endif
    _profile = [NSAccountDefaults.alloc initWithSuiteName:suitName];

    NSUserDefaults *ud = AppUserDefaultsShared();
    _token = ud.userToken;
}

#pragma mark -

@synthesize information = _information;
- (AccountEntity *)information {
    @synchronized(self) {
        if (_information) return _information;
        NSUserDefaults *ud = AppUserDefaultsShared();
        AccountEntity *ui = [AccountEntity.alloc initWithString:ud.AccountEntity error:nil];
        if (!ui) {
            ud.AccountEntity = nil;
            ui = AccountEntity.new;
        }
        _information = ui;
        return _information;
    }
}
- (void)setInformation:(AccountEntity *)information {
    @synchronized(self) {
        if (_information) {
            // 新获取的字段经常会是部分的，需要补全一下
            
        }
        
        _information = information;
        if (!information) {
            RFAssert(false, @"正常不会置空");
            return;
        }
        if (information.uid
            && self.uid != information.uid) {
            DebugLog(YES, @"MBUserInformationIDMismatch", @"用户信息 ID 不匹配");
            self.uid = information.uid;
            if (self.isCurrent) {
                AppUserDefaultsShared().lastUserID = self.uid;
            }
        }
        
        // 开始对接口/数据源取回的数据处理
        // 原则是保留能从用户信息接口获取的字段，如果是登录接口附加的信息则移动到 Account 上
        
        
#if DEBUG
        // 验证一些假设
        
#endif
        if (self.isCurrent) {
            AppUserDefaultsShared().AccountEntity = self.information.toJSONString;
        }
    }
}

#pragma mark -

- (void)updateInformationFromViewController:(UIViewController *)viewController complation:(MBGeneralCallback)complation {
    __block MBGeneralCallback callback = complation;
    if (!callback && viewController) {
        callback = ^(BOOL success, id _Nullable item, NSError *_Nullable error){
            if (error) {
                [AppHUD() alertError:error title:nil fallbackMessage:nil];
            }
        };
    }
    
    NSMutableDictionary *att = [NSMutableDictionary dictionaryWithCapacity:3];
    [API.global requestWithName:@"AcoountInfo" context:^(RFAPIRequestConext *c) {
        c.parameters = att;
        c.success = ^(id<RFAPITask>  _Nonnull task, AccountEntity *rsp) {
            self.hasLoginedThisSession = YES;
            self.information = rsp;

            if (self.isCurrent) {
                [AppEnv() setFlagOn:MBENVFlagUserInfoFetched];
            }
        };
        if (callback) {
            c.finished = ^(id<RFAPITask>  _Nullable task, BOOL success) {
                callback(success, self.information, task.error);
            };
        }
    }];
}

- (void)resetCookies {
    NSHTTPCookieStorage *cs = NSHTTPCookieStorage.sharedHTTPCookieStorage;
    [cs.cookies enumerateObjectsUsingBlock:^(NSHTTPCookie * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        [cs deleteCookie:obj];
    }];
}

@end

File: App/Model/Account/Account.swift
//
//  Account.swift
//  App
//

/**
 管理当前用户
 */
class Account: MBUser {
    // 有的项目登入时只返回认证信息，没有用户 ID，这时候需要用 userIDUndetermined 创建 Account 对象
    #if MBUserStringUID
    static let userIDUndetermined = "<undetermined>"
    #else
    static let userIDUndetermined = INT64_MAX
    #endif

    // MARK: - 状态

    override var description: String {
        "<Account \(ObjectIdentifier(self)): uid = \(uid), information: \(information.description), pofileFetched?: \(hasPofileFetchedThisSession)>"
    }

    /**
     用户基本信息

     不为空，操作上可以便捷一些
     */
    @objc var information: AccountEntity {
        get {
            objc_sync_enter(self)
            defer { objc_sync_exit(self) }

            if let ret = _information { return ret }
            var account = AppUserDefaultsShared().accountEntity
            if account == nil {
                AppUserDefaultsShared().accountEntity = nil
                account = AccountEntity()
            }
            _information = account
            return account!
        }
        set {
            objc_sync_enter(self)
            defer { objc_sync_exit(self) }

            _information = newValue

            #if MBUserStringUID
            let uidChanged = newValue.uid.length > 0 && uid != newValue.uid as String
            #else
            let uidChanged = newValue.uid > 0 && uid != newValue.uid
            #endif
            if uidChanged {
                if uid != Account.userIDUndetermined {
                    AppLog().critical("用户信息 ID 不匹配")
                }
                setValue(information.uid, forKeyPath: #keyPath(MBUser.uid))
            }

            // 开始对接口/数据源取回的数据处理
            // 原则是保留能从用户信息接口获取的字段
            // 如果是登录接口附加的信息则移动到 Account 上

            persistentInfomationToStore()
        }
    }
    private var _information: AccountEntity?
    private func persistentInfomationToStore() {
        guard isCurrent else { return }
        AppUserDefaultsShared().lastUserID = uid
        AppUserDefaultsShared().accountEntity = information
    }

    var token: String?

    var hasPofileFetchedThisSession: Bool = false

    // MARK: - 挂载

    @objc private(set) lazy var profile: NSAccountDefaults? = {
        let suitName = ("User\(uid)" as NSString).rf_MD5
        return NSAccountDefaults(suiteName: suitName)
    }()

    // MARK: - 流程

    /// 应用启动后初始流程
    class func setup() {
        precondition(AppUser() == nil, "应用初始化时应该还未设置当前用户")
        #if MBUserStringUID
        guard let userID = AppUserDefaultsShared().lastUserID else { return }
        #else
        let userID = AppUserDefaultsShared().lastUserID
        guard userID > 0 else { return }
        #endif
        guard let token = AppUserDefaultsShared().userToken else {
            AppLog().critical("Account has ID but no token")
            return
        }

        guard let user = Account(id: userID) else { fatalError() }
        user.token = token
        self.current = user
        user.updateInformation { c in
            c.failureCallback = APISlientFailureHandler(true)
        }
    }

    override class func onCurrentUserChanged(_ currentUser: MBUser?) {
        let user = currentUser as? Account
        let defaults = AppUserDefaultsShared()
        #if MBUserStringUID
        defaults.lastUserID = user?.uid
        #else
        defaults.lastUserID = user?.uid ?? 0
        #endif
        defaults.userToken = user?.token
        defaults.accountEntity = user?.information
        if !defaults.synchronize() {
            // 实际项目遇到过 UserDefaults 无法存储的 bug，需要用户重启设备才行
            // 处理方式可以参考： https://github.com/BB9z/iOS-Project-Template/blob/4.1/App/Model/Account/Account.swift#L123-L127
            NSLog("⚠️ 用户信息存储失败")
        }
    }

    override func onLogin() {
        debugPrint("当前用户 ID: \(uid), token: \(token ?? "null")")
        AppAPI().defineManager.authorizationHeader["token"] = token
        AppEnv().setFlagOn(.userHasLogged)
        if !hasPofileFetchedThisSession {
            updateInformation { c in
                c.failureCallback = APISlientFailureHandler(true)
            }
        }
    }
    override func onLogout() {
        AppEnv().setFlagOff(.userHasLogged)
        AppEnv().setFlagOff(.userInfoFetched)
        AppAPI().defineManager.authorizationHeader.removeObject(forKey: "token")
        profile?.synchronize()
    }

    /// 更新账号用户信息
    func updateInformation(requestContext context: (RFAPIRequestConext) -> Void) {
        API.requestName("AcoountInfo") { c in
            context(c)
            let inputSuccessCallback = c.successCallback
            c.success { task, rsp in
                guard let info = rsp as? AccountEntity else { fatalError() }
                if let cb = inputSuccessCallback {
                    cb(task, rsp)
                }
                self.hasPofileFetchedThisSession = true
                self.information = info
                if self.isCurrent {
                    AppEnv().setFlagOn(.userInfoFetched)
                }
            }
        }
    }
}
